File: fastApiProject/service/knowledgeFileService.py
from typing import List, Optional
from sqlalchemy.orm import Session

from entity.KnowledgeFile import KnowledgeFile
from database import KnowledgeFileModel


class KnowledgeFileService:
    def __init__(self):
        pass

    def create_knowledge_file(self, db: Session, knowledge_base_id: int, file_name: str,
                             file_size: int, file_hash: str, file_type: Optional[str] = None,
                             indexing_method: str = "semantic", status: str = "pending",
                             error_message: Optional[str] = None,
                             file_metadata: Optional[dict] = None) -> KnowledgeFile:
        """创建新的知识文件"""
        db_file = KnowledgeFileModel(
            knowledge_base_id=knowledge_base_id,
            file_name=file_name,
            file_size=file_size,
            file_type=file_type,
            file_hash=file_hash,
            indexing_method=indexing_method,
            status=status,
            error_message=error_message,
            file_metadata=file_metadata or {}
        )
        db.add(db_file)
        db.commit()
        db.refresh(db_file)

        return self._to_pydantic(db_file)

    def get_knowledge_file_by_id(self, db: Session, file_id: int) -> Optional[KnowledgeFile]:
        """根据ID获取知识文件"""
        db_file = db.query(KnowledgeFileModel).filter(
            KnowledgeFileModel.id == file_id,
            KnowledgeFileModel.is_deleted == 0
        ).first()

        if not db_file:
            return None

        return self._to_pydantic(db_file)

    def get_files_by_knowledge_base(self, db: Session, knowledge_base_id: int,
                                   skip: int = 0, limit: int = 100) -> List[KnowledgeFile]:
        """根据知识库ID获取文件列表"""
        db_files = db.query(KnowledgeFileModel).filter(
            KnowledgeFileModel.knowledge_base_id == knowledge_base_id,
            KnowledgeFileModel.is_deleted == 0
        ).order_by(KnowledgeFileModel.created_at.desc()).offset(skip).limit(limit).all()

        return [self._to_pydantic(f) for f in db_files]

    def update_file_status(self, db: Session, file_id: int, status: str,
                          error_message: Optional[str] = None) -> Optional[KnowledgeFile]:
        """更新文件状态"""
        db_file = db.query(KnowledgeFileModel).filter(
            KnowledgeFileModel.id == file_id,
            KnowledgeFileModel.is_deleted == 0
        ).first()

        if not db_file:
            return None

        db_file.status = status
        if error_message is not None:
            db_file.error_message = error_message
        db.commit()
        db.refresh(db_file)

        return self._to_pydantic(db_file)

    def update_file_metadata(self, db: Session, file_id: int,
                            file_metadata: dict) -> Optional[KnowledgeFile]:
        """更新文件元数据"""
        db_file = db.query(KnowledgeFileModel).filter(
            KnowledgeFileModel.id == file_id,
            KnowledgeFileModel.is_deleted == 0
        ).first()

        if not db_file:
            return None

        # 合并元数据
        if db_file.file_metadata:
            db_file.file_metadata = {**db_file.file_metadata, **file_metadata}
        else:
            db_file.file_metadata = file_metadata
        db.commit()
        db.refresh(db_file)

        return self._to_pydantic(db_file)

    def soft_delete_file(self, db: Session, file_id: int) -> bool:
        """软删除知识文件"""
        db_file = db.query(KnowledgeFileModel).filter(
            KnowledgeFileModel.id == file_id,
            KnowledgeFileModel.is_deleted == 0
        ).first()

        if not db_file:
            return False

        db_file.is_deleted = 1
        db.commit()
        return True

    def get_file_by_hash(self, db: Session, knowledge_base_id: int, file_hash: str) -> Optional[KnowledgeFile]:
        """根据文件哈希查找文件（用于去重）"""
        db_file = db.query(KnowledgeFileModel).filter(
            KnowledgeFileModel.knowledge_base_id == knowledge_base_id,
            KnowledgeFileModel.file_hash == file_hash,
            KnowledgeFileModel.is_deleted == 0
        ).first()

        if not db_file:
            return None

        return self._to_pydantic(db_file)

    def _to_pydantic(self, db_file: KnowledgeFileModel) -> KnowledgeFile:
        """ORM模型转Pydantic实体"""
        return KnowledgeFile(
            id=db_file.id,
            knowledge_base_id=db_file.knowledge_base_id,
            file_name=db_file.file_name,
            file_size=db_file.file_size,
            file_type=db_file.file_type,
            file_hash=db_file.file_hash,
            indexing_method=db_file.indexing_method,
            status=db_file.status,
            error_message=db_file.error_message,
            file_metadata=db_file.file_metadata,
            created_at=db_file.created_at,
            updated_at=db_file.updated_at,
            is_deleted=db_file.is_deleted
        )


# 全局KnowledgeFile服务实例
knowledge_file_service = KnowledgeFileService()

File: blog-agent/agent/src/main/java/com/xiafan/agent/service/KnowledgeFileService.java
package com.xiafan.agent.service;

import com.xiafan.agent.entity.KnowledgeFile;
import com.xiafan.agent.repository.KnowledgeChunkRepository;
import com.xiafan.agent.repository.KnowledgeFileRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;
import java.util.Optional;

/** Mirrors knowledgeFileService.py (+ the file indexing pipeline equivalent of process_file_task). */
@Service
public class KnowledgeFileService {

    private final KnowledgeFileRepository repository;
    private final KnowledgeChunkRepository chunkRepository;
    private final MinioService minioService;
    private final FileProcessingService fileProcessingService;
    private final KnowledgeChunkService chunkService;

    public KnowledgeFileService(KnowledgeFileRepository repository, KnowledgeChunkRepository chunkRepository,
                                MinioService minioService, FileProcessingService fileProcessingService,
                                KnowledgeChunkService chunkService) {
        this.repository = repository;
        this.chunkRepository = chunkRepository;
        this.minioService = minioService;
        this.fileProcessingService = fileProcessingService;
        this.chunkService = chunkService;
    }

    public KnowledgeFile createKnowledgeFile(int knowledgeBaseId, String fileName, int fileSize, String fileType,
                                             String fileHash, Map<String, Object> fileMetadata) {
        return repository.insert(knowledgeBaseId, fileName, fileSize, fileType, fileHash, fileMetadata);
    }

    public Optional<KnowledgeFile> getKnowledgeFileById(int fileId) {
        return repository.findById(fileId);
    }

    public List<KnowledgeFile> getFilesByKnowledgeBase(int kbId, int skip, int limit) {
        return repository.findByKnowledgeBase(kbId, skip, limit);
    }

    public Optional<KnowledgeFile> getFileByHash(int kbId, String fileHash) {
        return repository.findExistingByHash(kbId, fileHash);
    }

    public Optional<KnowledgeFile> updateFileStatus(int fileId, String status, String errorMessage) {
        if (repository.updateStatus(fileId, status, errorMessage) == 0) {
            return Optional.empty();
        }
        return repository.findById(fileId);
    }

    public Optional<KnowledgeFile> updateFileMetadata(int fileId, Map<String, Object> metadata) {
        if (repository.updateMetadata(fileId, metadata) == 0) {
            return Optional.empty();
        }
        return repository.findById(fileId);
    }

    public boolean softDeleteFile(int fileId) {
        return repository.softDelete(fileId) > 0;
    }

    /**
     * Background indexing task mirroring process_file_task: download from MinIO, extract text,
     * chunk, embed and persist chunks, then update the file status.
     */
    public void processFile(int fileId, String objectName, int knowledgeBaseId) {
        KnowledgeFile file = repository.findById(fileId).orElse(null);
        if (file == null) {
            return;
        }
        updateFileStatus(fileId, "processing", null);
        try {
            byte[] fileBytes = minioService.downloadFile(objectName);
            if (fileBytes == null) {
                throw new IllegalArgumentException("从MinIO下载文件失败: " + objectName);
            }
            String ext = MinioService.extOf(objectName);
            String content = fileProcessingService.extractText(fileBytes, ext);
            List<Map<String, Object>> chunksData = fileProcessingService.chunkText(
                    content, Map.of("file_id", fileId, "source", objectName));
            chunkService.batchCreateChunks(knowledgeBaseId, chunksData, fileId);
            updateFileStatus(fileId, "completed", null);
        } catch (Exception e) {
            updateFileStatus(fileId, "failed", e.getMessage());
        }
    }
}